mlflow/genai/utils/trace_utils.py

```diff
@@ -149,6 +149,41 @@ def resolve_outputs_from_trace(
     return outputs
 
 
+def resolve_conversation_from_session_traces(
+    session_traces: list[Trace],
+) -> list[dict[str, str]]:
+    """
+    Extract conversation history from session traces.
+
+    Args:
+        session_traces: List of traces from the same session, sorted by creation time.
+
+    Returns:
+        List of conversation messages in the format [{"role": "user"|"assistant", "content": str}].
+        Each trace contributes two messages: user (from input) and assistant (from output).
+    """
+    # Sort traces by creation time (timestamp_ms)
+    sorted_traces = sorted(session_traces, key=lambda t: t.info.timestamp_ms)
+
+    conversation = []
+    for trace in sorted_traces:
+        # Extract and parse input (user message)
+        inputs = extract_inputs_from_trace(trace)
+        if inputs is not None:
+            user_content = parse_inputs_to_str(inputs)
+            if user_content and user_content.strip():
+                conversation.append({"role": "user", "content": user_content})
```

Comment:
If an agent emits 2 messages in a row (ie with no user response), what is the trace input for the 2nd message? Will this cause problems in `resolve_conversation_from_session_trace`? One example of this in a chatbot would be "It looks like 5 minutes have elapsed. Do you still need help?"

Reply:
Removing the user turn makes sense